[DOCS] Add intro to traces (#1859)
grafana/tempo: docs/tempo/website/getting-started/tempo-in-grafana.md

---
title: Tempo in Grafana
weight: 400
---

# Tempo in Grafana

Grafana has a built-in Tempo datasource that can be used to query Tempo and visualize traces.  This page describes the high-level features and their availability.  Use the latest versions for best compatibility and stability.

## View trace by ID

The most basic functionality is to visualize a trace using its ID.  Select the Trace ID tab and enter the ID to view it. This functionality is enabled by default and is available in all versions of Grafana.
<p align="center"><img src="../assets/grafana-query.png" alt="View trace by ID"></p>

## Log search

Traces can be discovered by searching logs for entries containing trace IDs.  This is most useful when your application also logs relevant information about the trace that can also be searched, such as HTTP status code, customer ID, etc.  This feature requires Grafana 7.5 or later, with a linked Loki data source, and a [traceID derived field](https://grafana.com/docs/grafana/latest/datasources/loki/#derived-fields).

<p align="center"><img src="../assets/log-search.png" alt="Log Search"></p>

## Find traces using Tempo search

> Search for traces using common dimensions such as time range, duration, span tags, service names, etc. Use the trace view to quickly diagnose errors and high-latency events in your system.

Reviewer: ```suggestion
Search for traces using common dimensions such as time range, duration, span tags, service names, and more. Use the trace view to quickly diagnose errors and high-latency events in your system.
```

## View trace by ID
The most basic functionality is to visualize a trace using its ID.  Select the Trace ID tab and enter the ID to view it. This functionality is enabled by default and is available in all versions of Grafana.
<p align="center"><img src="../assets/grafana-query.png" alt="View trace by ID"></p>

## Log search
Traces can be discovered by searching logs for entries containing trace IDs.  This is most useful when your application also logs relevant information about the trace that can also be searched, such as HTTP status code, customer ID, etc.  This feature requires Grafana 7.5 or later, with a linked Loki data source, and a [traceID derived field](https://grafana.com/docs/grafana/latest/datasources/loki/#derived-fields).

<p align="center"><img src="../assets/log-search.png" alt="Log Search"></p>

## Find traces using Tempo search

> Search for traces using common dimensions such as time range, duration, span tags, service names, etc. Use the trace view to quickly diagnose errors and high latency events in your system.

Reviewer 2: hyphenate "high latency"?

Author: Yes, good catch.

> Search for traces using common dimensions such as time range, duration, span tags, service names, etc. Use the trace view to quickly diagnose errors and high latency events in your system.

Author: ```suggestion
Search for traces using common dimensions such as time range, duration, span tags, service names, etc. Use the trace view to quickly diagnose errors and high-latency events in your system.
```

For more information, refer to the [APM dashboard documentation]({{< relref "../metrics-generator/app-performance-mgmt/" >}}).

<p align="center"><img src="../assets/apm-overview.png" alt="APM dashboard overview"></p>

## Metrics from spans

> Requests Error Duration (RED) metrics can be used to drive service graphs and other ready-to-go visualizations of your span data. 

Reviewer 2: Should a hyperlink point to some RED content? Assuming there is some standard RED doc :slightly_smiling_face:

Author: I've updated this. Thank you for the suggestions.

## Metrics from spans

RED metrics can be used to drive service graphs and other ready-to-go visualizations of your span data. RED metrics represent:

- Rate, the number of requests per second
- Errors, the number of those requests that are failing
- Duration, the amount of time those requests take

For more information about RED method, refer to [The RED Method: How to instrument your services](https://grafana.com/blog/2018/08/02/the-red-method-how-to-instrument-your-services/).

>**Note:** Metrics generation is disabled by default. Contact Grafana Support to enable metrics generation in your organization. 

Once metrics-generator is enabled in your organization, refer to [Metrics-generator configuration]({{< relref "../configuration">}}) for information on metrics-generator options.

<p align="center"><img src="../assets/trace_service_graph.png" alt="Trace service graph"></p>

These metrics exist in your Hosted Metrics instance and can also be easily used to generate powerful custom dashboards.

<p align="center"><img src="../assets/trace_custom_metrics_dash.png" alt="Trace custom metrics dashboard"></p>

> Metrics automatically generate exemplars as well which allows easy metrics to trace linking. [Exemplars]({{< relref "exemplars" >}}) are GA in Grafana Cloud so you can also push your own.

Reviewer 3: ```suggestion
The metrics generator automatically generates exemplars as well which allows easy metrics to trace linking. [Exemplars]({{< relref "exemplars" >}}) are GA in Grafana Cloud so you can also push your own.
```
